src/udev/udev-builtin-net_id.c

```diff
@@ -243,17 +244,132 @@ static bool is_pci_ari_enabled(sd_device *dev) {
         return streq(a, "1");
 }
 
+static int pci_hotplug_slot_index(sd_device *dev) {
+        _cleanup_free_ char *slots = NULL;
+        const char *syspath;
+        _cleanup_(sd_device_unrefp) sd_device *pci = NULL;
+        _cleanup_closedir_ DIR *dir = NULL;
+        int r, hotplug_slot = 0;
+
+        if (!dev)
+                return 0;
+
+        /* ACPI _SUN — slot user number */
+        r = sd_device_new_from_subsystem_sysname(&pci, "subsystem", "pci");
+        if (r < 0)
+                return r;
+
+        r = sd_device_get_syspath(pci, &syspath);
+        if (r < 0)
+                return r;
+
+        slots = path_join(syspath, "slots");
+        if (!slots)
+                return -ENOMEM;
+
+        dir = opendir(slots);
+        if (!dir)
+                return -errno;
+
+        /* Walk up the PCI device tree and look for a match with some slot. Because slots
+           can be also associated with inner devices this might cause naming conflicts if more
+           leaf network devices are children of a same PCI bridge associated with the
+           single slot. We detect this issue later on and don't use slot based names
+           at all in such case. */
+        while (dev) {
+                struct dirent *dent;
+                const char *sysname;
+
+                if (sd_device_get_sysname(dev, &sysname) < 0)
+                        continue;
+
+                FOREACH_DIRENT_ALL(dent, dir, break) {
+                        unsigned i;
+                        _cleanup_free_ char *address = NULL, *str = NULL;
+
+                        if (dot_or_dot_dot(dent->d_name))
+                                continue;
+
+                        r = safe_atou_full(dent->d_name, 10, &i);
+                        if (r < 0 || i <= 0)
+                                continue;
+
+                        str = path_join(slots, dent->d_name, "address");
+                        if (!str)
+                                return -ENOMEM;
+
+                        /* match slot address with device by stripping the function */
+                        if (read_one_line_file(str, &address) >= 0 && startswith(sysname, address)) {
+                                hotplug_slot = i;
+                                break;
+                        }
+                }
+                if (hotplug_slot > 0)
+                        break;
+                if (sd_device_get_parent_with_subsystem_devtype(dev, "pci", NULL, &dev) < 0)
+                        break;
+
+                rewinddir(dir);
+        }
```

Comment:
I agree with @michich: this cannot work if slots are probed concurrently with this code. if slots appear in sysfs as our code here runs this will go wrong.

I am not convinced this is the right way to go if it's not reliable.